### PR TITLE
Add description for Property and commandline generation

### PR DESCRIPTION
**Pull request by tribal-tec**

### CODE REVIEW

File: brayns/common/PropertyMap.h [hunk 1]
```diff
@@ -30,199 +30,227 @@
...
-            Vec2i,
-            Vec2d,
-            Vec3i,
-            Vec3d,
-            Vec4d
-        };
-
-        template <typename T>
-        void assertValidType() const
+        UserInfo()
```
Comment (karjonas) [inline brayns/common/PropertyMap.h]: Why is it named UserInfo? Do you mean UsageInfo?

File: brayns/common/PropertyMap.h [hunk 2]
```diff
@@ -299,6 +327,14 @@ class PropertyMap
         throw std::runtime_error("No property found with name " + name);
     }
 
+    /** @return the user infofor the given property. */
```
Comment (karjonas) [inline brayns/common/PropertyMap.h]:
```suggestion
    /** @return the user info for the given property. */
```

File: brayns/common/PropertyMap.h (hunk 1 above)
Comment (tribal-tec) [reply brayns/common/PropertyMap.h]: Called it MetaData now

**Comment by tribal-tec** [comment]
ping

**Comment by tribal-tec** [comment]
Had another iteration where I added generic support for enums (with one minor limitation for integer enums), provided more docs & tests, and fix the casing (camelCase property to hyphenated-case for cmdline option).

**Comment by tribal-tec** [comment]
retest this please

File: brayns/common/utils/Utils.cpp [hunk 3]
```diff
@@ -90,4 +90,43 @@ std::string replaceFirstOccurrence(std::string input,
         return input;
     return input.replace(pos, toReplace.length(), replaceWith);
 }
+
```
Comment (hernando) [inline brayns/common/utils/Utils.cpp]: Shouldn't this file be lower case for consistency with other ones?

File: brayns/common/PropertyMap.h [hunk 4]
```diff
@@ -30,200 +30,236 @@
...
+ * Defines a single property which always has a name, a label for e.g. UIs
+ * and a typed value. Additionally, a user-defined minimum and maximum value
+ * range can be specified, otherwise the types' limits apply.
  */
-class PropertyMap
+struct Property
 {
-public:
-    PropertyMap() = default;
+    enum class Type
```
[the remark this answers is not in the document]
Comment (tribal-tec) [reply brayns/common/PropertyMap.h]: that's a different refactoring, no?

File: brayns/common/PropertyMap.h [hunk 5]
```diff
@@ -30,200 +30,236 @@
...
-            static_assert(std::is_same<T, int32_t>::value ||
-                              std::is_same<T, std::string>::value,
-                          "Invalid enum type.");
         }
-
-        template <typename T>
-        Type getType() const
+        MetaData(const std::string& label_, const std::string& description_)
+            : label(label_)
+            , description(description_)
```
[the remark this answers is not in the document]
Comment (tribal-tec) [reply brayns/common/PropertyMap.h]: I'd prefer having explicit, non-default arg constructors. But I can still do it.

File: brayns/common/PropertyMap.h [hunk 6]
```diff
@@ -30,200 +30,236 @@
...
+    {
+        const auto requestedType = getType<T>();
+        if (requestedType != type)
+            throw std::runtime_error("Type mismatch for property '" + name +
+                                     "'");
+    }
+    template <typename T>
+    T _castValue(const boost::any& v) const
+    {
+        _checkType<T>();
```
[the remark this answers is not in the document]
Comment (tribal-tec) [reply brayns/common/PropertyMap.h]: Don't know. I moved the code only, @karjonas was the original author of that.

File: brayns/common/commandlinePropertyMap.cpp [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+        case Property::Type::Vec2d:
+        case Property::Type::Vec3d:
+        case Property::Type::Vec4d:
+            valueSemantic = po::value<std::vector<double>>()->multitoken();
+            break;
+        }
+        if (valueSemantic)
+        {
+            const auto dashCaseName = camelCaseToSeparated(property->name, '-');
+            desc.add(boost::make_shared<po::option_description>(
```
[the remark this answers is not in the document]
Comment (tribal-tec) [reply brayns/common/commandlinePropertyMap.cpp]: no, boost wants its own shared_ptr type :)

File: brayns/common/commandlinePropertyMap.cpp [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+        case Property::Type::Vec3i:
+            valueSemantic = po::value<std::vector<int32_t>>()->multitoken();
+            break;
+        case Property::Type::Vec2d:
+        case Property::Type::Vec3d:
+        case Property::Type::Vec4d:
+            valueSemantic = po::value<std::vector<double>>()->multitoken();
+            break;
+        }
+        if (valueSemantic)
```
[the remark this answers is not in the document]
Comment (tribal-tec) [reply brayns/common/commandlinePropertyMap.cpp]: Actually never, unless someone adds a new property type and forgets it here.

File: plugins/VRPN/VRPNPlugin.cpp [hunk 9]
```diff
@@ -51,34 +51,32 @@ constexpr std::array<double, 4> to_array_4d(const vrpn_float64* quat)
     return {{quat[0], quat[1], quat[2], quat[3]}};
 }
 
-PropertyMap::Property getHeadPositionProperty()
+Property getHeadPositionProperty()
 {
-    PropertyMap::Property headPosition{HEAD_POSITION_PROP, "Head position",
-                                       HEAD_INIT_POS};
+    Property headPosition{HEAD_POSITION_PROP, HEAD_INIT_POS};
```
[the remark this answers is not in the document]
Comment (tribal-tec) [reply plugins/VRPN/VRPNPlugin.cpp]: yes, indeed

File: brayns/common/utils/Utils.cpp (hunk 3 above)
Comment (tribal-tec) [reply brayns/common/utils/Utils.cpp]: Yes, it should

File: brayns/common/commandlinePropertyMap.h [hunk 10]
```diff
@@ -0,0 +1,45 @@
+/* Copyright (c) 2015-2018, EPFL/Blue Brain Project
```
[the remark this answers is not in the document]
Comment (tribal-tec) [reply brayns/common/commandlinePropertyMap.h]: I can make it even a member function then.

File: brayns/common/commandlinePropertyMap.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (hernando) [reply brayns/common/commandlinePropertyMap.cpp]: Then replace the if with an assert

**Pull request merged**